Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,979 @@
+
+```python
+# no output (the function returns `None`) means that the test passed
+>>> test_demonstrating_the_given_decorator()
+```
+<!-- #endregion -->
+
+<!-- #region -->
+**Exploring other Core Strategies: Solution**
+
+Dictionaries of arbitrary size whose keys are positive-values integers and whose values are `True` or `False.
+
+```python
+>>> st.dictionaries(st.integers(min_value=1), st.booleans()).example()
+{110: True, 19091: True, 136348032: False, 78: False, 9877: False}
+```
+
+Length-4 strings whose elements are only lowercase vowels
+
+```python
+>>> st.text(alphabet="aeiou", min_size=4, max_size=4).example()
+'uiai'
+```
+
+Permutations of the list `[1, 2, 3, 4]`
+
+```python
+>>> st.permutations([1, 2, 3, 4]).example()
+[2, 3, 1, 4]
+```
+<!-- #endregion -->
+
+**Improving our tests using Hypothesis: Solution**
+
+Testing correctness by construction
+
+Write a hypothesis-driven test for the `count_vowels`; include this test in `test/test_basic_functions`.
+This is a test function where we can explicit construct a string in parts: its non-vowel characters, non-y vowels, and y-vowels.
+And thus, by constructing a string with a known number of vowel and non-vowel characters, we can know what the output of `count_vowels` *should* be for that input, and we can thus test for correctness in this way.
+We will want to read about the [st.text()](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.text) strategy to construct the different parts of the string.
+The standard library's built-in `string` module provides a string of all printable characters (`string.printable`).
+
+We should ask ourselves: How general are input strings that we are constructing? Are there regular patterns in the strings that might prevent our test from identifying edge case bugs in `count_vowels`?
+
+<!-- #region -->
+```python
+from string import printable
+from random import shuffle
+
+import hypothesis.strategies as st
+from hypothesis import given, note
+
+# a list of all printable non-vowel characters
+_not_vowels = "".join([l for l in printable if l.lower() not in set("aeiouy")])
+
+
+@given(
+    not_vowels=st.text(alphabet=_not_vowels),
+    vowels_but_not_ys=st.text(alphabet="aeiouAEIOU"),
+    ys=st.text(alphabet="yY"),
+)
+def test_count_vowels_hypothesis(not_vowels, vowels_but_not_ys, ys):
+    """
+    Constructs an input string with a known number of:
+       - non-vowel characters
+       - non-y vowel characters
+       - y characters
+    
+    and thus, by constructions, we can test that the output
+    of `count_vowels` agrees with the known number of vowels
+    """
+    # list of characters
+    letters = list(not_vowels) + list(vowels_but_not_ys) + list(ys)
+    
+    # We need to shuffle the ordering of our characters so that
+    # our input string isn't unnaturally patterned; e.g. always 
+    # have its vowels at the end
+    shuffle(letters)
+    in_string = "".join(letters)
```

Comment:
Trouble!  Hypothesis will seed the random-number generator, so you shouldn't show using `random` functions in PBTs.

Or, preferably, use the `st.random()` strategy to provide a `Random` instance that you use to do the shuffle, and explain why that's a good idea.